File: solutions/julia/satellite/1/satellite.jl
struct Tree
    root::String
    left::Union{Tree, Nothing}
    right::Union{Tree, Nothing}
end

function tree_from_traversals(preorder, inorder)
    !issetequal(preorder, inorder) &&
        throw(ArgumentError("inorder"))
    length(preorder) != length(unique(preorder)) &&
        throw(ArgumentError("preorder"))
    _gettree(preorder, inorder)
end

function _gettree(preorder, inorder)
    isempty(preorder) && return
    t = preorder[1]
    i = findfirst(==(t), inorder)
    l = _gettree(preorder[2:i], inorder[1:i-1])
    r = _gettree(preorder[i+1:end], inorder[i+1:end])
    Tree(t, l, r)
end

File: solutions/go/satellite/1/satellite.go
package satellite

import (
	"errors"
	"reflect"
)

type Node struct {
	Value string
	Left  *Node
	Right *Node
}

func TreeFromTraversals(preorder, inorder []string) (*Node, error) {
	if len(preorder) != len(inorder) {
		return nil, errors.New("traversals must have the same length")
	}

	preCounts := make(map[string]int, len(preorder))
	inCounts := make(map[string]int, len(inorder))
	for _, item := range preorder {
		preCounts[item]++
	}
	for _, item := range inorder {
		inCounts[item]++
	}

	if len(preCounts) != len(preorder) {
		return nil, errors.New("traversals must contain unique items")
	}

	if !reflect.DeepEqual(preCounts, inCounts) {
		return nil, errors.New("traversals must have the same elements")
	}

	var build func(pre, ino []string) *Node
	build = func(pre, ino []string) *Node {
		if len(pre) == 0 {
			return nil
		}

		value := pre[0]
		leftCount := -1
		for i, v := range ino {
			if v == value {
				leftCount = i
				break
			}
		}

		if leftCount == -1 {
			return nil
		}

		return &Node{
			Value: value,
			Left:  build(pre[1:leftCount+1], ino[:leftCount]),
			Right: build(pre[leftCount+1:], ino[leftCount+1:]),
		}
	}

	return build(preorder, inorder), nil
}